src/basic/async.c

```diff
@@ -79,29 +81,80 @@ int asynchronous_sync(pid_t *ret_pid) {
         return 0;
 }
 
-static void *close_thread(void *p) {
-        (void) pthread_setname_np(pthread_self(), "close");
+/* We encode the fd to close in the userdata pointer as an unsigned value. The highest bit indicates whether
+ * we need to fork again */
+#define NEED_DOUBLE_FORK (1U << (sizeof(unsigned) * 8 - 1))
 
-        assert_se(close_nointr(PTR_TO_FD(p)) != -EBADF);
-        return NULL;
+static int close_func(void *p) {
+        unsigned v = PTR_TO_UINT(p);
+
+        (void) prctl(PR_SET_NAME, (unsigned long*) "(close)");
+
+        /* Note: 💣 This function is invoked in a child process created via glibc's clone() wrapper. In such
+         *       children memory allocation is not allowed, since glibc does not release malloc mutexes in
+         *       clone() 💣 */
```

Comment:
Does this comment still valid even we drop threading from PID1?

Reply:
we use asynchronous_close() in logind too, not just pid1, and it might be others sooner or later, who knows. Hence better be safe than sorry, and keep the comment, maybe something adds threads assuming that clone()/clone3() isn't used and sooner or later and we might have a problem.